File: js/src/dsl.ts
// Tokenize figure-DSL source into statements. A faithful port of python/src/mtph/diagram/dsl.py.

export class DiagramSyntaxError extends Error {
  lineno: number;
  constructor(lineno: number, message: string) {
    super(`figure line ${lineno}: ${message}`);
    this.lineno = lineno;
  }
}

export interface Statement {
  command: string;
  positionals: string[];
  args: Record<string, string>;
  lineno: number;
}

const isSpace = (ch: string): boolean => /\s/.test(ch);

function tokenize(line: string, lineno: number): string[] {
  const tokens: string[] = [];
  let buf = "";
  let depth = 0;
  let inQuote = false;
  for (const ch of line) {
    if (inQuote) {
      buf += ch;
      if (ch === '"') inQuote = false;
      continue;
    }
    if (ch === '"') {
      inQuote = true;
      buf += ch;
    } else if (ch === "(") {
      depth++;
      buf += ch;
    } else if (ch === ")") {
      depth--;
      if (depth < 0) throw new DiagramSyntaxError(lineno, "unbalanced ')'");
      buf += ch;
    } else if (isSpace(ch) && depth === 0) {
      if (buf) {
        tokens.push(buf);
        buf = "";
      }
    } else {
      buf += ch;
    }
  }
  if (inQuote) throw new DiagramSyntaxError(lineno, "unterminated string (missing closing '\"')");
  if (depth !== 0) throw new DiagramSyntaxError(lineno, "unbalanced '(' in coordinate");
  if (buf) tokens.push(buf);
  return tokens;
}

export function parseDsl(source: string): Statement[] {
  const statements: Statement[] = [];
  const lines = source.split("\n");
  for (let idx = 0; idx < lines.length; idx++) {
    const line = lines[idx].trim();
    if (!line || line.startsWith("#")) continue;
    const tokens = tokenize(line, idx + 1);
    if (tokens.length === 0) continue;
    const stmt: Statement = { command: tokens[0], positionals: [], args: {}, lineno: idx + 1 };
    for (const tok of tokens.slice(1)) {
      const eq = tok.indexOf("=");
      if (eq > 0 && !tok.startsWith("(") && !tok.startsWith('"')) {
        stmt.args[tok.slice(0, eq)] = tok.slice(eq + 1);
      } else {
        stmt.positionals.push(tok);
      }
    }
    statements.push(stmt);
  }
  return statements;
}

File: python/src/mtph/diagram/dsl.py
"""Parse figure-DSL source into a list of :class:`Statement` objects.

Each non-empty, non-comment line is one statement::

    command [positional ...] key=value key=value ...

Values may be coordinates ``(x, y)`` (which contain spaces) or quoted strings, so we use a
small depth-aware tokenizer rather than ``str.split``. Interpretation of values (float vs
anchor vs string vs list) is left to the compiler.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List


class DiagramSyntaxError(ValueError):
    """A malformed figure DSL statement, with a 1-based line number."""

    def __init__(self, lineno: int, message: str):
        self.lineno = lineno
        super().__init__(f"figure line {lineno}: {message}")


@dataclass
class Statement:
    command: str
    positionals: List[str] = field(default_factory=list)
    args: Dict[str, str] = field(default_factory=dict)
    lineno: int = 0


def _tokenize(line: str, lineno: int) -> List[str]:
    tokens: List[str] = []
    buf = ""
    depth = 0  # parenthesis depth
    in_quote = False
    for ch in line:
        if in_quote:
            buf += ch
            if ch == '"':
                in_quote = False
            continue
        if ch == '"':
            in_quote = True
            buf += ch
        elif ch == "(":
            depth += 1
            buf += ch
        elif ch == ")":
            depth -= 1
            if depth < 0:
                raise DiagramSyntaxError(lineno, "unbalanced ')'")
            buf += ch
        elif ch.isspace() and depth == 0:
            if buf:
                tokens.append(buf)
                buf = ""
        else:
            buf += ch
    if in_quote:
        raise DiagramSyntaxError(lineno, "unterminated string (missing closing '\"')")
    if depth != 0:
        raise DiagramSyntaxError(lineno, "unbalanced '(' in coordinate")
    if buf:
        tokens.append(buf)
    return tokens


def parse_dsl(source: str) -> List[Statement]:
    statements: List[Statement] = []
    for idx, raw in enumerate(source.splitlines(), start=1):
        line = raw.strip()
        if not line or line.startswith("#"):
            continue
        tokens = _tokenize(line, idx)
        if not tokens:
            continue
        command = tokens[0]
        stmt = Statement(command=command, lineno=idx)
        for tok in tokens[1:]:
            # key=value, but not a coordinate like (1, 2) and not a quoted string
            eq = tok.find("=")
            if eq > 0 and not tok.startswith("(") and not tok.startswith('"'):
                key = tok[:eq]
                val = tok[eq + 1 :]
                stmt.args[key] = val
            else:
                stmt.positionals.append(tok)
        statements.append(stmt)
    return statements
